File: benchmark.py
"""
Massive multi-language benchmark runner.
Compiles and executes Go, C++, and Python implementations of the same
algorithms, measures compile time + execution time, and writes JSON results.

Usage:
    python3 benchmark.py                  # run all benchmarks, 3 runs each
    python3 benchmark.py --runs 5         # override run count
    python3 benchmark.py --filter sort    # only benchmarks matching name
    python3 benchmark.py --no-go          # skip Go
    python3 benchmark.py --no-cpp         # skip C++
    python3 benchmark.py --no-python      # skip Python
    python3 benchmark.py --timeout 120    # per-run timeout in seconds
"""

import argparse
import json
import os
import shutil
import stat
import subprocess
import sys
import time
import tempfile
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Optional


ROOT      = Path(__file__).parent
SRC       = ROOT / "tests"
GO_NAME   = "goSrc.go"
CPP_NAME  = "cppSrc.cpp"
PY_NAME   = "pythonSrc.py"
RESULTS   = ROOT / "results"
BUILD_DIR = ROOT / "build"
CPP_BIN   = shutil.which("g++") or "/usr/bin/g++"
GO_BIN    = shutil.which("go") or "/usr/lib/go/bin/go"


@dataclass
class RunResult:
    exit_code: int
    stdout: str
    stderr: str
    wall_seconds: float


@dataclass
class BenchmarkResult:
    name: str
    language: str
    compile_s: Optional[float]
    runs_s: list[float]        = field(default_factory=list)
    exec_mean_s: float         = 0.0
    exec_min_s: float          = 0.0
    exec_max_s: float          = 0.0
    total_s: float             = 0.0   # compile + exec_mean
    output: str                = ""
    error: str                 = ""
    skipped: bool              = False
    skip_reason: str           = ""


def _run(cmd: list[str], timeout: int, cwd: Optional[Path] = None) -> RunResult:
    t0 = time.perf_counter()
    try:
        p = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout,
            cwd=str(cwd) if cwd else None,
        )
        elapsed = time.perf_counter() - t0
        return RunResult(p.returncode, p.stdout.strip(), p.stderr.strip(), elapsed)
    except subprocess.TimeoutExpired:
        elapsed = time.perf_counter() - t0
        return RunResult(-1, "", f"TIMEOUT after {timeout}s", elapsed)
    except Exception as exc:
        elapsed = time.perf_counter() - t0
        return RunResult(-2, "", str(exc), elapsed)


def _compile_go(src: Path, out: Path, timeout: int) -> tuple[bool, float, str]:
    r = _run([GO_BIN, "build", "-o", str(out), str(src)], timeout)
    return r.exit_code == 0, r.wall_seconds, r.stderr


def _compile_cpp(src: Path, out: Path, timeout: int) -> tuple[bool, float, str]:
    r = _run([CPP_BIN, "-O2", "-std=c++17", "-pthread", "-o", str(out), str(src)], timeout)
    return r.exit_code == 0, r.wall_seconds, r.stderr


def run_benchmark(
    name: str,
    language: str,
    src: Path,
    n_runs: int,
    timeout: int,
) -> BenchmarkResult:
    res = BenchmarkResult(name=name, language=language, compile_s=None)

    binary: Optional[Path] = None

    if language == "go":
        if not src.exists():
            res.skipped, res.skip_reason = True, f"source not found: {src}"
            return res
        binary = BUILD_DIR / f"{name}_go"
        if os.name == "nt":
            binary = binary.with_suffix(".exe")
        ok, ctime, cerr = _compile_go(src, binary, timeout)
        res.compile_s = ctime
        if not ok:
            res.skipped, res.skip_reason = True, f"compile error:\n{cerr}"
            res.error = cerr
            return res

    elif language == "cpp":
        if not src.exists():
            res.skipped, res.skip_reason = True, f"source not found: {src}"
            return res
        binary = BUILD_DIR / f"{name}_cpp"
        if os.name == "nt":
            binary = binary.with_suffix(".exe")
        ok, ctime, cerr = _compile_cpp(src, binary, timeout)
        res.compile_s = ctime
        if not ok:
            res.skipped, res.skip_reason = True, f"compile error:\n{cerr}"
            res.error = cerr
            return res

    elif language == "python":
        if not src.exists():
            res.skipped, res.skip_reason = True, f"source not found: {src}"
            return res
        res.compile_s = 0.0   # interpreted

    for run_idx in range(n_runs):
        if language in ("go", "cpp"):
            cmd = [str(binary)]
        else:
            cmd = [sys.executable, str(src)]

        r = _run(cmd, timeout)

        if r.exit_code not in (0,):
            res.skipped = True
            res.skip_reason = f"run {run_idx+1} exited {r.exit_code}: {r.stderr}"
            res.error = r.stderr
            return res

        res.runs_s.append(r.wall_seconds)
        if run_idx == 0:
            res.output = r.stdout   # capture first run output

    if res.runs_s:
        res.exec_mean_s = sum(res.runs_s) / len(res.runs_s)
        res.exec_min_s  = min(res.runs_s)
        res.exec_max_s  = max(res.runs_s)
    res.total_s = (res.compile_s or 0.0) + res.exec_mean_s
    return res


@dataclass
class Benchmark:
    name: str
    srcPath: str
    category: str
    description: str

BENCHMARKS = [
    Benchmark("fibonacci_iter", "fibonacci",      "cpu",         "Fibonacci(40) iterative"),
    Benchmark("fibonacci_rec",  "fibRecursive",   "cpu",         "Fibonacci(42) recursive — exponential call tree"),
    Benchmark("prime_sieve",    "primeSieve",     "cpu",         "Sieve of Eratosthenes up to 5 million"),
    Benchmark("matrix_mul",     "matrixMultiply", "cpu",         "400x400 matrix multiplication (naive triple-loop)"),
    Benchmark("sort_ints",      "sorting",        "cpu",         "Sort 10 million integers (stdlib sort)"),
    Benchmark("json_roundtrip", "jsonParser",     "data",        "Encode + decode 200 000 objects as JSON"),
    Benchmark("string_build",   "stringBuilder",  "data",        "Build string from 500 000 formatted tokens"),
    Benchmark("hashmap",        "hashmap",        "data",        "Insert + read 2 million key-value pairs"),
    Benchmark("file_io",        "fileIO",         "io",          "Write then read 1 million lines to/from disk"),
    Benchmark("concurrency",    "concurrency",    "concurrency", "1 000 goroutines/threads doing arithmetic"),
]


RESET  = "\033[0m"
BOLD   = "\033[1m"
GREEN  = "\033[32m"
YELLOW = "\033[33m"
RED    = "\033[31m"
CYAN   = "\033[36m"
DIM    = "\033[2m"


def _color_time(t: float) -> str:
    if t < 0.1:
        return f"{GREEN}{t:.4f}s{RESET}"
    if t < 1.0:
        return f"{YELLOW}{t:.4f}s{RESET}"
    return f"{RED}{t:.4f}s{RESET}"


def print_result(r: BenchmarkResult):
    lang_pad = r.language.ljust(7)
    if r.skipped:
        print(f"  {DIM}{lang_pad}  SKIPPED — {r.skip_reason[:80]}{RESET}")
        return
    compile_str = f"compile={_color_time(r.compile_s)}" if r.compile_s is not None and r.language != "python" else f"compile={DIM}n/a{RESET}    "
    print(f"  {CYAN}{lang_pad}{RESET}  {compile_str}  "
          f"exec_mean={_color_time(r.exec_mean_s)}  "
          f"exec_min={_color_time(r.exec_min_s)}  "
          f"total={_color_time(r.total_s)}")

def pars_args():
    parser = argparse.ArgumentParser(description="Multi-language benchmark runner")
    parser.add_argument("--runs",      type=int, default=3,   help="exec runs per benchmark (default 3)")
    parser.add_argument("--timeout",   type=int, default=180, help="per-run timeout seconds (default 180)")
    parser.add_argument("--filter",    type=str, default="",  help="only run benchmarks whose name contains this string")
    parser.add_argument("--no-go",     action="store_true",   help="skip Go")
    parser.add_argument("--no-cpp",    action="store_true",   help="skip C++")
    parser.add_argument("--no-python", action="store_true",   help="skip Python")
    parser.add_argument("--output",    type=str, default="",  help="path to write JSON results (default: results/run_<timestamp>.json)")
    parser.add_argument("--cpp-compiler", type=str, default="",  help="path to C++ compiler (default: g++)")
    parser.add_argument("--go-compiler", type=str, default="",  help="path to Go compiler (default: go)")
    return parser.parse_args()

def main():
    args = pars_args()

    BUILD_DIR.mkdir(parents=True, exist_ok=True)
    RESULTS.mkdir(parents=True, exist_ok=True)

    languages = []
    if not args.no_go:
        languages.append("go")
        if args.go_compiler:
            global GO_BIN
            GO_BIN = args.go_compiler
    if not args.no_cpp:
        languages.append("cpp")
        if args.cpp_compiler:
            global CPP_BIN
            CPP_BIN = args.cpp_compiler
    if not args.no_python:
        languages.append("python")

    benchmarks = BENCHMARKS
    if args.filter:
        benchmarks = [b for b in benchmarks if args.filter.lower() in b.name.lower()]

    print(f"\n{BOLD}═══════════════════════════════════════════════════════{RESET}")
    print(f"{BOLD}  Multi-language Benchmark Suite{RESET}")
    print(f"{BOLD}═══════════════════════════════════════════════════════{RESET}")
    print(f"  Languages : {', '.join(languages)}")
    print(f"  Benchmarks: {len(benchmarks)}")
    print(f"  Runs each : {args.runs}")
    print(f"  Timeout   : {args.timeout}s per run")
    print()

    all_results: list[BenchmarkResult] = []
    total_start = time.perf_counter()

    for benchmark in benchmarks:
        src_dest = SRC / benchmark.srcPath
        src_map = { "go": (src_dest / GO_NAME), "cpp": (src_dest / CPP_NAME), "python": (src_dest / PY_NAME) }

        print(f"{BOLD}[{benchmark.category.upper()}] {benchmark.name}{RESET}  {DIM}{benchmark.description}{RESET}")

        for lang in languages:
            src = src_map[lang]
            r = run_benchmark(benchmark.name, lang, src, args.runs, args.timeout)
            all_results.append(r)
            print_result(r)

        print()

    elapsed_total = time.perf_counter() - total_start

    print(f"\n{BOLD}═══════════════════════════════════════════════════════{RESET}")
    print(f"{BOLD}  Summary — execution time (mean), normalized to Python{RESET}")
    print(f"{BOLD}═══════════════════════════════════════════════════════{RESET}")

    header = f"  {'Benchmark':<22} {'Python':>10} {'Go':>10} {'C++':>10}  {'Go speedup':>12}  {'C++ speedup':>12}"
    print(header)
    print("  " + "─" * (len(header) - 2))

    for benchmark in benchmarks:
        row: dict[str, BenchmarkResult] = {
            r.language: r for r in all_results if r.name == benchmark.name
        }
        py_t  = row.get("python")
        go_t  = row.get("go")
        cpp_t = row.get("cpp")

        def fmt(r: Optional[BenchmarkResult]) -> str:
            if r is None or r.skipped: return "  skip"
            return f"{r.exec_mean_s:>8.3f}s"

        def speedup(base: Optional[BenchmarkResult], target: Optional[BenchmarkResult]) -> str:
            if base is None or target is None:
                return "     n/a"
            if base.skipped or target.skipped:
                return "     n/a"
            if target.exec_mean_s == 0:
                return "     ∞"
            ratio = base.exec_mean_s / target.exec_mean_s
            return f"{ratio:>8.2f}x"

        print(f"  {benchmark.name:<22} {fmt(py_t)} {fmt(go_t)} {fmt(cpp_t)}  "
              f"{speedup(py_t, go_t):>12}  {speedup(py_t, cpp_t):>12}")

    print(f"\n  Total wall time: {elapsed_total:.1f}s")

    ts = int(time.time())
    out_path = Path(args.output) if args.output else RESULTS / f"run_{ts}.json"

    payload = {
        "timestamp":          ts,
        "runs_per_benchmark": args.runs,
        "timeout_s":          args.timeout,
        "total_wall_s":       round(elapsed_total, 3),
        "results":            [asdict(r) for r in all_results],
    }
    out_path.write_text(json.dumps(payload, indent=2))
    print(f"\n  Results written → {out_path}\n")


if __name__ == "__main__":
    main()

File: benchmark.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"math"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"time"
)

var (
	ROOT, _   = os.Getwd()
	SRC       = filepath.Join(ROOT, "tests")
	RESULTS   = filepath.Join(ROOT, "results")
	BUILD_DIR = filepath.Join(ROOT, "build")
)

var (
	runs        = flag.Int   ("runs"        , 3    , "Number of runs per test")
	timeout     = flag.Int   ("timeout"     , 180  , "Time after test is being stoped")
	filter      = flag.String("filter"      , ""   , "Filter all tests maching this value")
	noGo        = flag.Bool  ("no-go"       , false, "Filter Go tests")
	noCpp       = flag.Bool  ("no-cpp"      , false, "Filter C++ tests")
	noPy        = flag.Bool  ("no-python"   , false, "Filter Python tests")
	output      = flag.String("output"      , ""   , "Set the output path")
	cppCompiler = flag.String("cpp-compiler", "g++", "Provide c++ bin path")
	goCompiler  = flag.String("go-compiler" , "go" , "Provide go bin path")
	noColor     = flag.Bool  ("no-color"    , false, "disable ANSI colors")
)

const (
    esc   = "\033["
    reset = esc + "0m"
    bold  = esc + "1m"
    dim   = esc + "2m"

    fgBlack   = esc + "30m"
    fgRed     = esc + "31m"
    fgGreen   = esc + "32m"
    fgYellow  = esc + "33m"
    fgBlue    = esc + "34m"
    fgMagenta = esc + "35m"
    fgCyan    = esc + "36m"
    fgWhite   = esc + "37m"

    fgBrightBlack   = esc + "90m"
    fgBrightRed     = esc + "91m"
    fgBrightGreen   = esc + "92m"
    fgBrightYellow  = esc + "93m"
    fgBrightBlue    = esc + "94m"
    fgBrightMagenta = esc + "95m"
    fgBrightCyan    = esc + "96m"
    fgBrightWhite   = esc + "97m"

    bgBlue = esc + "44m"

    clearLine = "\r" + esc + "2K"
)

var colorEnabled = true

func c(codes ...string) string {
	if !colorEnabled {
		return ""
	}
	return strings.Join(codes, "")
}

func colored(text string, codes ...string) string {
	if !colorEnabled {
		return text
	}
	return strings.Join(codes, "") + text + reset
}

func langColor(lang string) string {
	switch lang {
        case "go":
            return fgBrightCyan
        case "c++":
            return fgBrightBlue
        case "python":
            return fgBrightYellow
        default:
            return fgBrightMagenta
	}
}

func langTag(lang string) string {
	tag := fmt.Sprintf(" %-7s", lang)
	return colored(tag, bold, langColor(lang))
}

func categoryColor(cat string) string {
	switch cat {
        case "cpu":
            return fgBrightRed
        case "data":
            return fgBrightGreen
        case "io":
            return fgBrightYellow
        case "concurrency":
            return fgBrightMagenta
        default:
            return fgBrightWhite
	}
}

func fmtTime(s float64) string {
	var text string
	var col string
	switch {
        case s <= 0:
            return colored("  —      ", dim)
        case s < 0.001:
            text = fmt.Sprintf("%7.3fms", s*1000)
            col = fgBrightGreen
        case s < 0.1:
            text = fmt.Sprintf("%7.2fms", s*1000)
            col = fgBrightGreen
        case s < 1.0:
            text = fmt.Sprintf("%7.0fms", s*1000)
            col = fgBrightYellow
        case s < 10:
            text = fmt.Sprintf("%7.3f s", s)
            col = fgYellow
        default:
            text = fmt.Sprintf("%7.2f s", s)
            col = fgRed
	}
	return colored(text, col)
}

const barWidth = 28

func progressBar(done, total int, label string) string {
	pct := float64(done) / float64(total)
	filled := int(pct * barWidth)
	if filled > barWidth {
		filled = barWidth
	}
	empty := barWidth - filled

	bar := colored("[", dim) +
		colored(strings.Repeat("█", filled), fgBrightCyan) +
		colored(strings.Repeat("░", empty), fgBrightBlack) +
		colored("]", dim)

	pctStr := colored(fmt.Sprintf("%3.0f%%", pct*100), bold, fgBrightWhite)
	labelStr := colored(fmt.Sprintf(" %-28s", truncate(label, 28)), fgBrightBlack)

	return fmt.Sprintf("%s %s%s", bar, pctStr, labelStr)
}

func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	return s[:max-1] + "…"
}

func printProgress(done, total int, label string) {
	fmt.Print(clearLine + progressBar(done, total, label))
}

var spinFrames = []string{"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"}

func spinnerFrame(i int) string {
	return colored(spinFrames[i%len(spinFrames)], fgBrightCyan)
}

func printBenchHeader(idx, total int, b Benchmark) {
	catCol := categoryColor(b.Category)
	catTag := colored(fmt.Sprintf(" %-12s", strings.ToUpper(b.Category)), bold, catCol)
	nameStr := colored(b.Name, bold, fgBrightWhite)
	descStr := colored("  "+b.Description, dim)
	progress := colored(fmt.Sprintf("[%d/%d]", idx+1, total), fgBrightBlack)

	fmt.Printf("\n%s%s %s%s %s\n", catTag, reset, nameStr, descStr, progress)
	fmt.Println(colored(strings.Repeat("─", 64), fgBrightBlack))
}

func printResultRow(res TestResult) {
	tag := langTag(res.Language)

	if res.Skipped {
		reason := truncate(res.SkipReason, 48)
		fmt.Printf("  %s  %s %s\n", tag, colored("SKIPPED", bold, fgBrightBlack), colored(reason, dim))
		return
	}

	compileStr := ""
	if res.CompileS != nil && *res.CompileS > 0 {
		compileStr = colored("compile ", dim) + fmtTime(*res.CompileS) + "  "
	} else if res.CompileS != nil {
		compileStr = colored("compile   interp  ", dim) + "  "
	}

	execStr := colored("exec ", dim) +
		fmtTime(res.ExecMeanS) +
		colored(fmt.Sprintf("  (min %s  max %s)", fmtTime(res.ExecMinS), fmtTime(res.ExecMaxS)), dim)

	fmt.Printf("  %s  %s%s\n", tag, compileStr, execStr)
}

func printRunProgress(run, total int, lang string) {
	dots := strings.Repeat(colored("●", fgBrightCyan), run) + strings.Repeat(colored("○", fgBrightBlack), total-run)
	label := colored(fmt.Sprintf(" run %d/%d of %s", run, total, colored(lang, langColor(lang))), fgBrightBlack)
	fmt.Printf("%s  %s%s", clearLine, dots, label)
}

func printSummaryRow(br BenchmarkResult) {
    catCol := categoryColor(br.Category)
    catStr := colored(fmt.Sprintf("  %-12s", br.Category), catCol)

    marker := " "
    if br.Internal {
        marker = colored("⊙", fgBrightMagenta)
    }
    nameStr := fmt.Sprintf("%s %-28s", marker, truncate(br.Name, 27))

    times := map[string]float64{}
    for i, r := range br.Results {
        if !r.Skipped {
            lang := r.Language
            if br.Internal {
                lang += strconv.Itoa(i + 1)
            }
            times[lang] = r.ExecMeanS
        }
    }

    best := math.MaxFloat64
    for _, t := range times {
        if t < best {
            best = t
        }
    }

    fmtCell := func(lang string) string {
        t, ok := times[lang]
        if !ok {
            return colored("    —    ", fgBrightBlack)
        }
        s := fmtTime(t)
        if t == best {
            return colored("★", fgBrightYellow) + s
        }
        return " " + s
    }

    outFmt := ""
    if br.Internal {
        for i := range br.Results {
            outFmt += fmtCell("go" + strconv.Itoa(i + 1)) + "  "
        }
    } else {
        outFmt = fmt.Sprintf("%s  %s  %s", fmtCell("go"), fmtCell("c++"), fmtCell("python"))

    }

    fmt.Printf(
        "%s  %s  %s\n",
        catStr,
        nameStr,
        outFmt,
    )
}

func printSummary(results []BenchmarkResult, totalWall float64) {
	fmt.Printf("\n%s\n", colored(strings.Repeat("═", 72), fgBrightBlack))
	fmt.Printf("%s\n\n", colored("  RESULTS", bold, fgBrightWhite))

	fmt.Printf("%s\n", colored("  MULTILANG", bold, fgBrightWhite))
	fmt.Printf(
        "  %-15s    %-28s                   %s   %s  %s\n",
		colored("CATEGORY    ", bold, fgBrightBlack),
		colored("BENCHMARK", bold, fgBrightBlack),
		colored("    GO   ", bold, fgBrightCyan),
		colored("   C++   ", bold, fgBrightBlue),
		colored(" PYTHON  ", bold, fgBrightYellow),
	)
	fmt.Println(colored("  "+strings.Repeat("─", 68), fgBrightBlack))

	for _, br := range results {
        if !br.Internal {
            printSummaryRow(br)
        }
    }

	fmt.Printf("\n%s\n", colored("  INTERNAL", bold, fgBrightWhite))
    fmt.Printf(
        "  %-15s    %-28s                   %s\n",
		colored("CATEGORY    ", bold, fgBrightBlack),
		colored("BENCHMARK", bold, fgBrightBlack),
		colored("   GOs...", bold, fgBrightCyan),
	)
	fmt.Println(colored("  "+strings.Repeat("─", 68), fgBrightBlack))

	for _, br := range results {
        if br.Internal {
            printSummaryRow(br)
        }
    }

    fmt.Println(colored("  "+strings.Repeat("─", 68), fgBrightBlack))
    fmt.Printf(
        "  %s %s\n\n",
        colored("total wall time", dim),
        colored(fmt.Sprintf("%.2fs", totalWall), bold, fgBrightWhite),
    )
}

type RunResult struct {
    ExitCode    int     `json:"exit_code"`
    Stdout      string  `json:"stdout"`
    Stderr      string  `json:"stderr"`
    WallSecond  float64 `json:"wall_seconds"`
}

type TestResult struct {
    TestName   string    `json:"test_name"`
    Language   string    `json:"language"`
    CompileS   *float64  `json:"compile_s"`
    RunsS      []float64 `json:"runs_s"`
    ExecMeanS  float64   `json:"exec_mean_s"`
    ExecMinS   float64   `json:"exec_min_s"`
    ExecMaxS   float64   `json:"exec_max_s"`
    TotalS     float64   `json:"total_s"`
    Output     string    `json:"output"`
    Error      string    `json:"error"`
    Skipped    bool      `json:"skipped"`
    SkipReason string    `json:"skip_reason"`
}

type BenchmarkResult struct {
    Name     string       `json:"name"`
    Internal bool         `json:"internal"`
    Category string       `json:"category"`
    Results  []TestResult `json:"results"`
}

type Benchmark struct {
    Name        string  `json:"name"`
    SrcPath     string  `json:"src_path"`
    Category    string  `json:"category"`
    Description string  `json:"description"`
    Internal    bool    `json:"internal"`
}

func runCmd(cwd string, cmd string, args ...string) RunResult {
    start := time.Now()
    ctx, cancel := context.WithTimeout(context.Background(), time.Duration(*timeout)*time.Second)
    defer cancel()

    c := exec.CommandContext(ctx, cmd, args...)
    if cwd != "" {
        c.Dir = cwd
    }

    out, err := c.Output()
    elapsed := time.Since(start).Seconds()

    if ctx.Err() == context.DeadlineExceeded {
        return RunResult{-1, "", fmt.Sprintf("TIMEOUT after %ds", timeout), elapsed}
    }

    if err != nil {
        if ee, ok := err.(*exec.ExitError); ok {
            return RunResult{ee.ExitCode(), string(out), string(ee.Stderr), elapsed}
        }
        return RunResult{-2, "", err.Error(), elapsed}
    }

    return RunResult{0, strings.TrimSpace(string(out)), "", elapsed}
}

func compileGo(src, out string) (float64, string, bool) {
    r := runCmd("", *goCompiler, "build", "-o", out, src)
    return r.WallSecond, r.Stderr, r.ExitCode == 0
}

func compileCpp(src, out string) (float64, string, bool) {
    r := runCmd("", *cppCompiler, "-O2", "-std=c++17", "-pthread", "-o", out, src)
    return r.WallSecond, r.Stderr, r.ExitCode == 0
}

func buildTestSources(testSrc string) ([]string, error) {
	var srcs []string
	srcBase := filepath.Join(SRC, testSrc)
	err := filepath.Walk(
        srcBase,
        func(path string, f os.FileInfo, err error) error {
            if filepath.Ext(path) != "" {
                srcs = append(srcs, path)
            }
            return err
        })
    return srcs, err
}

func skipLang(lang string) bool {
    switch lang {
        case "go":
            return *noGo
        case "c++":
            return *noCpp
        case "python":
            return *noPy
    }
    return false
}

func mapExt2Lang(path string) string {
    switch filepath.Ext(path) {
        case ".go":
            return "go"
        case ".cpp":
            return "c++"
        case ".py":
            return "python"
        }
    return "unknown"
}

func runLangBenchmark(benchName, lang, src string) TestResult {
    var binary string
    sum := 0.0

    testName := strings.TrimSuffix(filepath.Base(src), filepath.Ext(src))
    res := TestResult{TestName: testName, Language: lang}

    if skipLang(lang) {
        res.Skipped = true
        res.SkipReason = "filtered lang: " + lang
        goto done
    }

    if _, err := os.Stat(src); err != nil {
        res.Skipped = true
        res.SkipReason = "source not found: " + src
        goto done
    }

    switch lang {
        case "go":
            binary = filepath.Join(BUILD_DIR, benchName + "_go")
            if runtime.GOOS == "windows" {
                binary += ".exe"
            }
            fmt.Printf("%s  %s  %s compiling…",
                clearLine, langTag(lang),
                colored(testName, fgBrightBlack))

            ct, ce, ok := compileGo(src, binary)
            res.CompileS = &ct
            if !ok {
                fmt.Print(clearLine)
                res.Skipped, res.Error, res.SkipReason = true, ce, ce
                return res
            }

        case "c++":
            binary = filepath.Join(BUILD_DIR, benchName + "_cpp")
            if runtime.GOOS == "windows" {
                binary += ".exe"
            }
            fmt.Printf("%s  %s  %s compiling…",
                clearLine, langTag(lang),
                colored(testName, fgBrightBlack))

            ct, ce, ok := compileCpp(src, binary)
            res.CompileS = &ct
            if !ok {
                fmt.Print(clearLine)
                res.Skipped, res.Error, res.SkipReason = true, ce, ce
                goto done
            }

        default:
            v := 0.0
            res.CompileS = &v
    }

    for i := 0; i < *runs; i++ {
        printRunProgress(i, *runs, lang)

        var r RunResult
        if lang == "python" {
            r = runCmd("", "python", src)
        } else {
            r = runCmd("", binary)
        }

        if r.ExitCode != 0 {
            fmt.Print(clearLine)
            res.Skipped, res.Error, res.SkipReason = true, r.Stderr, r.Stderr
            goto done
        }

        res.RunsS = append(res.RunsS, r.WallSecond)
        if i == 0 {
            res.Output = r.Stdout
        }
    }
    fmt.Print(clearLine)

    res.ExecMinS = math.MaxFloat64
    for _, t := range res.RunsS {
        sum += t
        if t < res.ExecMinS {
            res.ExecMinS = t
        }
        if t > res.ExecMaxS {
            res.ExecMaxS = t
        }
    }
    if len(res.RunsS) > 0 {
        res.ExecMeanS = sum / float64(len(res.RunsS))
    }
    res.TotalS = *res.CompileS + res.ExecMeanS

done:
    printResultRow(res)
    return res
}

func runBenchmark(b Benchmark) BenchmarkResult {
    res := BenchmarkResult{
        Name:     b.Name,
        Internal: b.Internal,
        Category: b.Category,
        Results:  make([]TestResult, 0),
    }

    srcs, err := buildTestSources(b.SrcPath)
    if err != nil {
        fmt.Printf("  %s reading %s: %v\n", colored("ERROR", bold, fgBrightRed), b.Name, err)
        return res
    }

    for _, src := range srcs {
        lang := mapExt2Lang(src)
        r := runLangBenchmark(b.Name, lang, src)
        res.Results = append(res.Results, r)
    }

    return res
}

func init() {
    flag.Parse()
    colorEnabled = !*noColor

    _ = os.MkdirAll(BUILD_DIR, 0755)
    _ = os.MkdirAll(RESULTS, 0755)
}

func loadBenchmarks() ([]Benchmark, error) {
    var benchmarks []Benchmark
    data, err := os.ReadFile("benchmarks.json")
    if err != nil {
        goto errLoading
    }
    err = json.Unmarshal(data, &benchmarks)
    if err != nil {
        goto errLoading
    }
    return benchmarks, nil
errLoading:
    return nil, err
}

func main() {
    benchmarks, err := loadBenchmarks()
    if err != nil {
        fmt.Printf("%s loading benchmarks: %v\n", colored("ERROR", bold, fgBrightRed), err)
        return
    }

    var active []Benchmark
    for _, b := range benchmarks {
        if *filter == "" || strings.Contains(strings.ToLower(b.Name), strings.ToLower(*filter)) {
            active = append(active, b)
        }
    }

    total := len(active)

    fmt.Printf("\n%s\n", colored("  ▶  benchmark suite", bold, fgBrightWhite))
    fmt.Printf("  %s %s   %s %s   %s %s\n\n",
        colored("runs", dim), colored(fmt.Sprintf("%d", *runs), fgBrightCyan),
        colored("timeout", dim), colored(fmt.Sprintf("%ds", *timeout), fgBrightCyan),
        colored("benchmarks", dim), colored(fmt.Sprintf("%d", total), fgBrightCyan),
    )

    start := time.Now()
    var results []BenchmarkResult

    for i, b := range active {
        printProgress(i, total, b.Name)
        fmt.Println()

        printBenchHeader(i, total, b)

        r := runBenchmark(b)
        results = append(results, r)
    }

    printProgress(total, total, "done")
    fmt.Println()

    wallTime := time.Since(start).Seconds()

    printSummary(results, wallTime)

    payload := map[string]any{
        "timestamp":           time.Now().Unix(),
        "runs_per_benchmark":  *runs,
        "timeout_s":           *timeout,
        "total_wall_s":        wallTime,
        "benchmarks":          results,
    }

    data, _ := json.MarshalIndent(payload, "", "  ")

    outPath := *output
    if outPath == "" {
        outPath = filepath.Join(RESULTS, fmt.Sprintf("run_%d.json", time.Now().Unix()))
    }

    _ = os.WriteFile(outPath, data, 0644)

    fmt.Printf("  %s %s\n\n",
        colored("results →", dim),
        colored(outPath, fgBrightCyan),
    )
}
